### PR TITLE
Handle $_ in mbe

### LINKED ISSUE #6926: `$_` is rejected in MBE (gtk-rs clone!)
Minimized code:

```rust
macro_rules! clone {
    ($_:ident) => {};
}

fn main() {
    clone!(foo);
}
```

### PR DESCRIPTION
**Pull request by lnicola**
Fixes #6926

### CODE REVIEW

**Comment by lnicola** [comment]
I'm pretty sure this isn't the right solution, though...

File: crates/mbe/src/mbe_expander/transcriber.rs [hunk 1]
```diff
@@ -115,7 +115,9 @@ fn expand_subtree(
     ExpandResult { value: tt::Subtree { delimiter: template.delimiter, token_trees: tts }, err }
 }
 
-fn expand_var(ctx: &mut ExpandCtx, v: &SmolStr) -> ExpandResult<Fragment> {
+fn expand_var(ctx: &mut ExpandCtx, v: Option<&SmolStr>) -> ExpandResult<Fragment> {
+    let underscore = SmolStr::new_inline("_");
+    let v = v.unwrap_or(&underscore);
```
Comment (lnicola) [inline crates/mbe/src/mbe_expander/transcriber.rs:120]: This is all awkward because `name: &'a SmolStr` so we can't make a new one when needed.

**Comment by lnicola** [comment]
And apparently `$__:tt` ends up as `Ident`...

**Comment by edwin0cheng** [comment]
> I'm pretty sure this isn't the right solution, though...

Yeah, I am pretty sure we didn't handle underscore as special case in mbe. So let me check what's happened in this week end?

**Comment by edwin0cheng** [comment]
Oh, I understand why now. 

https://github.com/rust-analyzer/rust-analyzer/blob/067067a6c11bb5afda98f5af14bfdec4744e7812/crates/parser/src/syntax_kind/generated.rs#L269-L279

We treat `underscore` as punct . But why it is a punct ? It should not be right ?

**Comment by edwin0cheng** [comment]
If we have to make it to punct for parser, I recommend the following changes:

https://github.com/rust-analyzer/rust-analyzer/blob/7f3ba7d57fd2f43f3cfc012f898a26d940bf5cc8/crates/mbe/src/syntax_bridge.rs#L316

Change to 
```rust
result.push(if k.is_punct() && k != UNDERSCORE {
```

and 

https://github.com/rust-analyzer/rust-analyzer/blob/7f3ba7d57fd2f43f3cfc012f898a26d940bf5cc8/crates/mbe/src/syntax_bridge.rs#L380-L381

Add:
```rust
IDENT => make_leaf!(Ident),
UNDERSCORE => make_leaf!(Ident),   // <---
k if k.is_keyword() => make_leaf!(Ident),
```

**Comment by lnicola** [comment]
Thanks for the pointers, that doesn't seem to break anything else.

**Comment by edwin0cheng** [comment]
LGTM, bors r+

**Pull request merged**